snowplow-common-internal/kinesis/src/main/scala/com.snowplowanalytics.snowplow/sources/KinesisSource.scala

```diff
@@ -0,0 +1,252 @@
+ * You may obtain a copy of the Snowplow Community License Version 1.0 at https://docs.snowplow.io/community-license-1.0
+ */
+package com.snowplowanalytics.snowplow.sources
+
+import cats._
+import cats.implicits._
+
+import cats.effect.{Async, Resource, Sync}
+
+import eu.timepit.refined.types.all.PosInt
+
+import fs2.Stream
+import fs2.aws.kinesis.{CommittableRecord, Kinesis, KinesisConsumerSettings}
+
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import software.amazon.awssdk.regions.providers.DefaultAwsRegionProviderChain
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient
+
+import scala.collection.mutable.ArrayBuffer
+
+import java.net.{InetAddress, URI}
+import java.util.concurrent.Semaphore
+import java.util.{Date, UUID}
+
+// kinesis
+import software.amazon.kinesis.common.{ConfigsBuilder, InitialPositionInStream, InitialPositionInStreamExtended}
+import software.amazon.kinesis.coordinator.Scheduler
+import software.amazon.kinesis.processor.SingleStreamTracker
+import software.amazon.kinesis.exceptions.ShutdownException
+import software.amazon.kinesis.metrics.MetricsLevel
+import software.amazon.kinesis.processor.ShardRecordProcessorFactory
+import software.amazon.kinesis.retrieval.fanout.FanOutConfig
+import software.amazon.kinesis.retrieval.polling.PollingConfig
+
+// snowplow
+import com.snowplowanalytics.snowplow.sources.internal.{Checkpointer, LowLevelEvents, LowLevelSource}
+
+object KinesisSource {
+
+  private implicit def logger[F[_]: Sync] = Slf4jLogger.getLogger[F]
+
+  def build[F[_]: Parallel: Async](config: KinesisSourceConfig): SourceAndAck[F] =
+    LowLevelSource.toSourceAndAck(lowLevel(config))
+
+  private type KinesisCheckpointer[F[_]] = Checkpointer[F, List[KinesisMetadata[F]]]
+
+  private def toMetadata[F[_]: Sync]: CommittableRecord => KinesisMetadata[F] = cr =>
+    KinesisMetadata(cr.shardId, cr.sequenceNumber, cr.isLastInShard, cr.lastRecordSemaphore, cr.checkpoint)
+
+  final case class KinesisMetadata[F[_]](
+    shardId: String,
+    sequenceNumber: String,
+    isLastInShard: Boolean,
+    lastRecordSemaphore: Semaphore,
+    ack: F[Unit]
+  )
+
+  private def lowLevel[F[_]: Parallel: Async](config: KinesisSourceConfig): LowLevelSource[F, List[KinesisMetadata[F]]] =
+    new LowLevelSource[F, List[KinesisMetadata[F]]] {
+      def checkpointer: KinesisCheckpointer[F] = kinesisCheckpointer[F]
+
+      def stream: Stream[F, Stream[F, LowLevelEvents[List[KinesisMetadata[F]]]]] =
+        Stream.emit(kinesisStream(config))
+    }
+
+  private def kinesisCheckpointer[F[_]: Parallel: Sync]: KinesisCheckpointer[F] = new KinesisCheckpointer[F] {
+    def combine(x: List[KinesisMetadata[F]], y: List[KinesisMetadata[F]]): List[KinesisMetadata[F]] = x ::: y
```

Comment:
Here goes my biggest criticism of the entire PR.....

Here, your checkpointer is accumulating a `List[KinesisMetadata[F]]`, which means you are storing a `KinesisMetadata[F]` in memory for every single event (until we checkpoint and clear the list).

For our applications which work in windows (lake loader, BRA, rdb transformer) there might be *millions* of events processed in between checkpoints.  So the list might grow to millions long.

It's not a huge problem -- we probably won't run out of memory -- but it's also completely unnecessary.  We only need to store in memory one `KinesisMetadata[F]` _per shard_.  Indeed, you later [drop and ignore the majority of the list](https://github.com/snowplow-incubator/snowplow-lake-loader/blob/e979ba15671e6ba3ca87e12280bb8701aeba2077/snowplow-common-internal/kinesis/src/main/scala/com.snowplowanalytics.snowplow/sources/KinesisSource.scala#L84).

How about, instead of

```scala
type KinesisCheckpointer[F[_]] = Checkpointer[F, List[KinesisMetadata[F]]]
```

...try to build it using

```scala
type KinesisCheckpointer[F[_]] = Checkpointer[F, Map[String, KinesisMetadata[F]]]
```

...where the map keys are the shard IDs.  This is approximately [what the RDB transformer does](https://github.com/snowplow/snowplow-rdb-loader/blob/5.7.3/modules/transformer-kinesis/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/stream/kinesis/KinesisCheckpointer.scala#L23), although in other ways the transformer implementation is slightly different.